### PR TITLE
[12.0][ADD] supplier_calendar

### PR DESCRIPTION
**Pull request by NuriaXifre**
New module that adds the possibility to assign a calendar to a supplier. In purchase orders, only working days of the supplier will be taken into account when computing order and scheduled dates.
~ @LoisRForgeFlow @JordiBForgeFlow @ForgeFlow

### CODE REVIEW

File: supplier_calendar/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2020 ForgeFlow
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Supplier Calendar",
+    "summary": "Supplier Calendar",
+    'version': '12.0.1.0.0',
+    "category": "Purchase Management",
+    "author": "ForgeFlow",
+    "contributors": [
```
Comment (legalsylvain) [inline supplier_calendar/__manifest__.py]: did you meant ``maintainers`` : https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L13

File: supplier_calendar/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2020 ForgeFlow
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Supplier Calendar",
+    "summary": "Supplier Calendar",
+    'version': '12.0.1.0.0',
+    "category": "Purchase Management",
+    "author": "ForgeFlow",
```
Comment (legalsylvain) [inline supplier_calendar/__manifest__.py]: Add ``Odoo Community Association (OCA)`` key.

File: supplier_calendar/models/product.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (legalsylvain) [inline supplier_calendar/models/product.py]: maybe too much lines ;-)

File: supplier_calendar/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (legalsylvain) [inline supplier_calendar/models/res_partner.py]: also here.

File: supplier_calendar/models/stock_rule.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+from odoo import models, api, fields
+from datetime import datetime
+
+class StockRule(models.Model):
+    _inherit = 'stock.rule'
+
+    def _get_purchase_order_date(self, product_id, product_qty, product_uom,
+                                 values, partner, schedule_date):
+        res = super(StockRule, self)._get_purchase_order_date(product_id,
```
Comment (legalsylvain) [inline supplier_calendar/models/stock_rule.py]: non blocking point : you can replace by : 

``res = super()._get_purchase_order_date(product_id,``

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Some comments inline.
I'll test on runbot and complete by a functional review.

File: supplier_calendar/README.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+~~~~~~~
+
+* ForgeFlow
+
+Contributors
+~~~~~~~~~~~~
+
+* Núria Martín <nuria.martin@forgeflow.com>
+* Jordi Ballester <jordi.ballester@forgeflow.com>
+* Lois Rilo <lois.rilo@eforgeflow>
```
Comment (LoisRForgeFlow) [inline supplier_calendar/README.rst:89]: typo here

File: supplier_calendar/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    "summary": "Supplier Calendar",
+    'version': '12.0.1.0.0',
+    "category": "Purchase Management",
+    "author": "ForgeFlow,"
+               "Odoo Community Association (OCA)",
+    "maintainers": [
+        'Nuria Martin <nuria.martin@forgeflow.com>',
+        'Jordi Ballester <jordi.ballester@forgeflow.com>',
+        'Lois Rilo <lois.rilo@forgeflow.com>'
+    ],
```
Comment (LoisRForgeFlow) [inline supplier_calendar/__manifest__.py]: mainteiners key is used by ocabot to allow the **github users** specified to merge even if they are not PSC. See this link for more reference: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/oca_module_lifecycle_maintainer_role.rst#maintainer-role-and-responsibilities

So, in short, this should be something like `"maintainers": ["NuriaMForgeFlow"]`

File: supplier_calendar/models/__init__.py [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (LoisRForgeFlow) [inline supplier_calendar/models/__init__.py]: agpl stil here, but you can just remove it in this case, copyrights are usually not neede in init files.

File: supplier_calendar/tests/test_supplier_calendar.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        wednesday = fields.Date.to_date('2097-01-09 09:00:00')
+        self.assertEqual(date_order, wednesday)  # Wednesday
+
+    def test_02_purchase_order_supplier_calendar_global_leaves(self):
+        # Global leaves
+        self.calendar.write({
+            'global_leave_ids': [
+                (0, False, {
+                    'name': 'Test',
+                    'date_from': '2097-01-15',  # Thursday
```
Comment (LoisRForgeFlow) [inline supplier_calendar/tests/test_supplier_calendar.py]: Tuesday

File: supplier_calendar/tests/test_supplier_calendar.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                    'date_to': '2097-01-18',  # Friday
+                }),
+            ],
+        })
+
+        reference = "2097-01-14 9:00:00"  # Monday ·counts as day one
```
Comment (LoisRForgeFlow) [inline supplier_calendar/tests/test_supplier_calendar.py]: "Monday ·counts as day one" - why?

File: supplier_calendar/tests/test_supplier_calendar.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        # With calendar
+        result = self.company_partner.supplier_plan_days(reference, 3).date()
+        next_thursday = fields.Date.to_date("2097-01-22")
```
Comment (LoisRForgeFlow) [inline supplier_calendar/tests/test_supplier_calendar.py]: tuesday

**Review by LoisRForgeFlow — COMMENTED** [review]
Code review, some comments inline:

File: supplier_calendar/tests/test_supplier_calendar.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            'global_leave_ids': [
+                (0, False, {
+                    'name': 'Test',
+                    'date_from': '2097-01-14',  # Monday
+                    'date_to': '2097-01-19',  # Saturday
+                }),
+            ],
+        })
+
+        reference = "2097-01-14 09:00:00"  # Monday
```
Comment (LoisRForgeFlow) [inline supplier_calendar/tests/test_supplier_calendar.py:91]: I was not asking to remove the comment, I was wondering wether it should be like this or not, kind of opening the discussion :thinking:

File: supplier_calendar/tests/test_supplier_calendar.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        # With calendar
+        result = self.company_partner.supplier_plan_days(reference, 3).date()
+        next_wednesday = fields.Date.to_date("2097-01-23")
```
Comment (LoisRForgeFlow) [inline supplier_calendar/tests/test_supplier_calendar.py:94]: So, extending my previous comment, my concert is here. If LT is 3 days, i would expect it to be Thursday here not Wednesday.

Comment (LoisRForgeFlow) [reply supplier_calendar/tests/test_supplier_calendar.py:94]: cc @JordiBForgeFlow

Comment (NuriaXifre) [reply supplier_calendar/tests/test_supplier_calendar.py:94]: I understand your suggestion

**Review by JordiBForgeFlow — APPROVED** [review]

**Review by LoisRForgeFlow — APPROVED** [review]
question solved! Thanks! LGTM

**Pull request merged**